src/ifcopenshell-python/ifcopenshell/util/unit.py

```diff
@@ -624,3 +622,83 @@ def format_length(
     elif unit_system == "metric":
         rounded_val = round(value / precision) * precision
         return f"{rounded_val:.{decimal_places}f}"
+
+
+def get_base_type_name(
+        content_type: ifcopenshell.ifcopenshell_wrapper.named_type | ifcopenshell.ifcopenshell_wrapper.type_declaration) -> ifcopenshell.ifcopenshell_wrapper.type_declaration | None:
+    cur_decl = content_type
+    while hasattr(cur_decl, "declared_type") is True:
+        cur_decl = cur_decl.declared_type()
+        if hasattr(cur_decl, "name") is False:
+            continue
+        if cur_decl.name() == "IfcLengthMeasure":
+            return cur_decl
+
+    if isinstance(cur_decl, ifcopenshell.ifcopenshell_wrapper.aggregation_type):
+        res = cur_decl.type_of_element()
+        cur_decl = res.declared_type()
+        if hasattr(cur_decl, "name") and cur_decl.name() == "IfcLengthMeasure":
+            return cur_decl
+        while hasattr(cur_decl, "declared_type") is True:
+            cur_decl = cur_decl.declared_type()
+            if hasattr(cur_decl, "name") is False:
+                continue
+            if cur_decl.name() == "IfcLengthMeasure":
+                return cur_decl
+
+    return None
+
+
+def convert_file_units(ifc_file: ifcopenshell.file, target_units: str) -> ifcopenshell.file:
+    """Converts all units in an IFC file to the specified target units. Returns a new file."""
+    prefix = "MILLI" if target_units == "MILLIMETERS" else None
+
+    file_patched = ifcopenshell.api.run("project.create_file", version=ifc_file.schema)
+    if ifc_file.schema == "IFC2X3":
+        user = file_patched.add(ifc_file.by_type("IfcProject")[0].OwnerHistory.OwningUser)
+        application = file_patched.add(ifc_file.by_type("IfcProject")[0].OwnerHistory.OwningApplication)
+        old_get_user = ifcopenshell.api.owner.settings.get_user
+        old_get_application = ifcopenshell.api.owner.settings.get_application
+        ifcopenshell.api.owner.settings.get_user = lambda ifc: user
+        ifcopenshell.api.owner.settings.get_application = lambda ifc: application
```

Comment:
Is this still needed?

Reply:
I just naively copied the IFC2x3 stuff from the patch not really knowing its intended purpose. 

I am fine with removing it

Reply:
I just removed it in https://github.com/IfcOpenShell/IfcOpenShell/pull/4125/commits/f5500da0af1ad79a3350bd304bf7b03365b065ea

If anyone wants the IFC2x3 functionality we can add it to the patch instead. Or suggest using the conversion utility to convert from 2x3 to a more recent version :)